### PR TITLE
[SPARK-28144][SPARK-29294][SS] Upgrade Kafka to 2.4.0

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

This patch upgrades the version of Kafka to 2.4, which supports Scala 2.13.

There're some incompatible changes in Kafka 2.4 which the patch addresses as well:

* `ZkUtils` is removed -> Replaced with `KafkaZkClient`
* Majority of methods are removed in `AdminUtils` -> Replaced with `AdminZkClient`
* Method signature of `Scheduler.schedule` is changed (return type) -> leverage `DeterministicScheduler` to avoid implementing `ScheduledFuture`

### Why are the changes needed?

* Kafka 2.4 supports Scala 2.13

### Does this PR introduce any user-facing change?

No, as Kafka API is known to be compatible across versions.

### How was this patch tested?

Existing UTs

### CODE REVIEW

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala [hunk 1]
```diff
@@ -575,15 +577,12 @@ class KafkaTestUtils(
...
-    assert(
-      !zkUtils.pathExists(getDeleteTopicPath(topic)),
-      s"${getDeleteTopicPath(topic)} still exists")
-    assert(!zkUtils.pathExists(getTopicPath(topic)), s"${getTopicPath(topic)} still exists")
+    assert(!zkClient.isTopicMarkedForDeletion(topic), "topic is still marked for deletion")
+    assert(!zkClient.topicExists(topic), "topic still exists")
     // ensure that the topic-partition has been deleted from all brokers' replica managers
     assert(servers.forall(server => topicAndPartitions.forall(tp =>
-      server.replicaManager.getPartition(tp) == None)),
+      server.replicaManager.getPartition(tp) == HostedPartition.None)),
```
Comment (HeartSaVioR) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala]: The return type is changed - it no longer returns Option, but includes None in its return type.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala [hunk 2]
```diff
@@ -626,9 +624,9 @@ class KafkaTestUtils(
     def isPropagated = server.dataPlaneRequestProcessor.metadataCache
         .getPartitionInfo(topic, partition) match {
       case Some(partitionState) =>
-        zkUtils.getLeaderForPartition(topic, partition).isDefined &&
-          Request.isValidBrokerId(partitionState.basePartitionState.leader) &&
-          !partitionState.basePartitionState.replicas.isEmpty
+        zkClient.getLeaderForPartition(new TopicPartition(topic, partition)).isDefined &&
+          Request.isValidBrokerId(partitionState.leader) &&
```
Comment (HeartSaVioR) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala]: partitionState.basePartitionState no longer exists. partitionState.basePartitionState.XXX seem to be moved to partitionState.XXX, at least for these we use

File: external/kafka-0-10/pom.xml [hunk 3]
```diff
@@ -111,6 +111,11 @@
       <groupId>org.apache.spark</groupId>
       <artifactId>spark-tags_${scala.binary.version}</artifactId>
     </dependency>
+    <dependency>
+      <groupId>org.jmock</groupId>
+      <artifactId>jmock-junit4</artifactId>
```
Comment (HeartSaVioR) [inline external/kafka-0-10/pom.xml]: This is to leverage DeterministicScheduler - scope is test.

File: external/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/mocks/MockScheduler.scala [hunk 4]
```diff
@@ -41,57 +42,33 @@ import org.apache.kafka.common.utils.Time
...
+  val scheduler = new DeterministicScheduler()
 
   def isStarted: Boolean = true
 
   def startup(): Unit = {}
 
   def shutdown(): Unit = synchronized {
-    tasks.foreach(_.fun())
-    tasks.clear()
+    scheduler.runUntilIdle()
```
Comment (HeartSaVioR) [inline external/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/mocks/MockScheduler.scala]: DeterministicScheduler doesn't support shutdown, so we don't call it here.

**Comment by HeartSaVioR** [comment]
cc. @srowen @dongjoon-hyun @gaborgsomogyi

**Comment by HeartSaVioR** [comment]
retest this, please

File: external/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/mocks/MockScheduler.scala [hunk 5]
```diff
@@ -41,57 +42,33 @@ import org.apache.kafka.common.utils.Time
...
-    var nextExecution: Long,
-    val period: Long) extends Ordered[MockTask] {
-  def periodic: Boolean = period >= 0
-  def compare(t: MockTask): Int = {
-    java.lang.Long.compare(t.nextExecution, nextExecution)
+      unit: TimeUnit = TimeUnit.MILLISECONDS): ScheduledFuture[_] = synchronized {
+    val runnable = new Runnable {
+      override def run(): Unit = fun()
+    }
+    if(period >= 0) {
```
Comment (srowen) [inline external/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/mocks/MockScheduler.scala]: Nit: space after if

File: external/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/mocks/MockScheduler.scala [hunk 6]
```diff
@@ -41,57 +42,33 @@ import org.apache.kafka.common.utils.Time
...
+      scheduler.scheduleAtFixedRate(runnable, delay, period, unit)
```
Comment (srowen) [inline external/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/mocks/MockScheduler.scala]: If it takes a Runnable, can you just pass `fun`? If the SAM rules allow it it should pass it as an implementation of Runnable. no big deal of course

**Review by srowen — COMMENTED** [review]
The change looks OK pending tests.
What implication does this have for compatibility with Kafka brokers? We probably need to do this eventually anyway, and better in Spark 3, regardless.

File: external/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/mocks/MockScheduler.scala (hunk 6 above)
Comment (HeartSaVioR) [reply external/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/mocks/MockScheduler.scala]: It doesn't allow to just pass `fun`, but allow `() => fun()` so will use it.

**Comment by HeartSaVioR** [comment]
> What implication does this have for compatibility with Kafka brokers?

Kafka community hasn't been mentioning compatibility with Kafka brokers across versions, as they have broad backward compatibility if I understand correctly. They're versioning the protocols.

File: external/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/mocks/MockScheduler.scala (hunk 6 above)
Comment (HeartSaVioR) [reply external/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/mocks/MockScheduler.scala]: `schedule` has overloaded methods which reject `() => fun()`. Just reverted back.

**Comment by dongjoon-hyun** [comment]
Thank you so much, @HeartSaVioR !

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala (hunk 2 above)
Comment (dongjoon-hyun) [reply external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaTestUtils.scala]: Got it.

File: external/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/KafkaTestUtils.scala [hunk 7]
```diff
@@ -85,19 +87,27 @@ private[kafka010] class KafkaTestUtils extends Logging {
...
+
   // Set up the Embedded Zookeeper server and get the proper Zookeeper port
   private def setupEmbeddedZookeeper(): Unit = {
     // Zookeeper server startup
     zookeeper = new EmbeddedZookeeper(s"$zkHost:$zkPort")
     // Get the actual zookeeper binding port
     zkPort = zookeeper.actualPort
-    zkUtils = ZkUtils(s"$zkHost:$zkPort", zkSessionTimeout, zkConnectionTimeout, false)
+    zkClient = KafkaZkClient(s"$zkHost:$zkPort", isSecure = false, zkSessionTimeout,
+      zkConnectionTimeout, 1, new SystemTime())
```
Comment (dongjoon-hyun) [inline external/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/KafkaTestUtils.scala]: `new SystemTime()` -> `Time.SYSTEM`?

File: external/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/KafkaTestUtils.scala [hunk 8]
```diff
@@ -27,13 +27,14 @@ import scala.annotation.tailrec
...
-import kafka.admin.AdminUtils
 import kafka.api.Request
 import kafka.server.{KafkaConfig, KafkaServer}
-import kafka.utils.ZkUtils
+import kafka.zk.{AdminZkClient, KafkaZkClient}
 import org.apache.kafka.clients.producer.{KafkaProducer, ProducerRecord}
+import org.apache.kafka.common.TopicPartition
 import org.apache.kafka.common.network.ListenerName
 import org.apache.kafka.common.serialization.StringSerializer
+import org.apache.kafka.common.utils.SystemTime
```
Comment (dongjoon-hyun) [inline external/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/KafkaTestUtils.scala]: `import org.apache.kafka.common.utils.Time`?

File: external/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/mocks/MockScheduler.scala [hunk 9]
```diff
@@ -41,57 +42,33 @@ import org.apache.kafka.common.utils.Time
...
       delay: Long = 0,
       period: Long = -1,
-      unit: TimeUnit = TimeUnit.MILLISECONDS): Unit = synchronized {
-    tasks += MockTask(name, fun, time.milliseconds + delay, period = period)
-    tick()
-  }
-
-}
-
-case class MockTask(
```
Comment (dongjoon-hyun) [inline external/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/mocks/MockScheduler.scala]: Thank you for removing this.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.
Since my minor comments are not important for this, I'll merge this to master.
Thank you, @HeartSaVioR and @srowen .

**Comment by HeartSaVioR** [comment]
Thanks all for reviewing and merging!

**Comment by HeartSaVioR** [comment]
I just addressed these review comments in #26979 as I'm seeing benefit on the change. Please take a look. Thanks in advance!

**Comment by gaborgsomogyi** [comment]
Thanks guys to take care.

**Comment by dongjoon-hyun** [comment]
Finally, you are back, @gaborgsomogyi . :)

**Comment by gaborgsomogyi** [comment]
Happy New Year @dongjoon-hyun ! Thanks for your warm welcome :)
After a relatively long pause hope I can continue to help the community...

**Comment by HyukjinKwon** [comment]
Everybody went for holidays :-)

**Comment by gaborgsomogyi** [comment]
My "holiday" was bit forced because of my health but hope now it will be better...

**Comment by gaborgsomogyi** [comment]
Aaaand of course same best wishes to everybody :)

**Comment by HyukjinKwon** [comment]
Oh .. I see.. Happy to see you back :D.

**Pull request closed without merging**